src/js/node/http2.ts

```diff
@@ -350,348 +1297,1498 @@ const constants = {
+  HTTP2_HEADER_HOST,
+  HTTP2_HEADER_IF_MATCH,
+  HTTP2_HEADER_IF_MODIFIED_SINCE,
+  HTTP2_HEADER_IF_NONE_MATCH,
+  HTTP2_HEADER_IF_RANGE,
+  HTTP2_HEADER_IF_UNMODIFIED_SINCE,
+  HTTP2_HEADER_LAST_MODIFIED,
+  HTTP2_HEADER_LOCATION,
+  HTTP2_HEADER_MAX_FORWARDS,
+  HTTP2_HEADER_PROXY_AUTHORIZATION,
+  HTTP2_HEADER_RANGE,
+  HTTP2_HEADER_REFERER,
+  HTTP2_HEADER_RETRY_AFTER,
+  HTTP2_HEADER_TK,
+  HTTP2_HEADER_UPGRADE_INSECURE_REQUESTS,
+  HTTP2_HEADER_USER_AGENT,
+  HTTP2_HEADER_X_CONTENT_TYPE_OPTIONS,
+]);
+
+function assertValidPseudoHeader(key) {
+  if (!kValidPseudoHeaders.has(key)) {
+    throw new ERR_HTTP2_INVALID_PSEUDOHEADER(key);
+  }
+}
+hideFromStack(assertValidPseudoHeader);
+
+const NoPayloadMethods = new Set([HTTP2_METHOD_DELETE, HTTP2_METHOD_GET, HTTP2_METHOD_HEAD]);
+
+type Settings = {
+  headerTableSize: number;
+  enablePush: boolean;
+  maxConcurrentStreams: number;
+  initialWindowSize: number;
+  maxFrameSize: number;
+  maxHeaderListSize: number;
+  maxHeaderSize: number;
+};
+
+class Http2Session extends EventEmitter {}
+
+function streamErrorFromCode(code: number) {
+  const error = new Error(`Stream closed with error code ${nameForErrorCode[code] || code}`);
+  error.code = "ERR_HTTP2_STREAM_ERROR";
+  error.errno = code;
+  return error;
+}
+hideFromStack(streamErrorFromCode);
+function sessionErrorFromCode(code: number) {
+  const error = new Error(`Session closed with error code ${nameForErrorCode[code] || code}`);
+  error.code = "ERR_HTTP2_SESSION_ERROR";
+  error.errno = code;
+  return error;
+}
+hideFromStack(sessionErrorFromCode);
+
+function assertSession(session) {
+  if (!session) {
+    const error = new Error(`ERR_HTTP2_INVALID_SESSION: The session has been destroyed`);
+    error.code = "ERR_HTTP2_INVALID_SESSION";
+    throw error;
+  }
+}
+hideFromStack(assertSession);
+
+function pushToStream(stream, data) {
+  // if (stream.writableEnded) return;
+  const queue = stream[bunHTTP2StreamReadQueue];
+  if (queue.isEmpty()) {
+    if (stream.push(data)) return;
+  }
+  queue.push(data);
+}
+
+enum StreamState {
+  EndedCalled = 1 << 0, // 00001 = 1
+  WantTrailer = 1 << 1, // 00010 = 2
+  FinalCalled = 1 << 2, // 00100 = 4
+  Closed = 1 << 3, // 01000 = 8
+  StreamResponded = 1 << 4, // 10000 = 16
+  WritableClosed = 1 << 5, // 100000 = 32
+}
+function markWritableDone(stream: Http2Stream) {
+  const _final = stream[bunHTTP2StreamFinal];
+  if (typeof _final === "function") {
+    stream[bunHTTP2StreamFinal] = null;
+    _final();
+    stream[bunHTTP2StreamStatus] |= StreamState.WritableClosed | StreamState.FinalCalled;
+    return;
+  }
+  stream[bunHTTP2StreamStatus] |= StreamState.WritableClosed;
+}
+function markStreamClosed(stream: Http2Stream) {
+  const status = stream[bunHTTP2StreamStatus];
+
+  if ((status & StreamState.Closed) === 0) {
+    stream[bunHTTP2StreamStatus] = status | StreamState.Closed;
+    markWritableDone(stream);
+  }
+}
+
+class Http2Stream extends Duplex {
```

Comment:
Is this a class in Node?

Reply:
It is https://github.com/nodejs/node/blob/4dc76a990f0037401b6bf8c982d7401c2533f23b/lib/internal/http2/core.js#L1957